### PR TITLE
NR-23881 - Give users possibility to add MDC to logging

### PR DESCRIPTION
**Pull request by StanlieK**
# Include MDC data into logs forwarded to NR from log4j-2 and logback

### Overview
This PR will add support for including MDC data in logs forwarded to NR from log4j-2 and logback. It maps the MDC map into log events prefixed by `mdc.`. E.g. if you have the MDC key `example`, it will have the key `mdc.example` in New Relic.

The whole feature is **disabled** by default and can be enabled by setting:
* `application_logging.forwarding.include_mdc.enabled` to `true`
* `NEW_RELIC_APPLICATION_LOGGING_FORWARDING_INCLUDE_MDC_ENABLED` to `true`

**Important:** This feature is not available for log4j 1.x as the agent does not support forwarding of the logs at all.

#### Example - Logback with disabled INCLUDE_MDC feature
Screenshot 2022-05-27 at 15 55 04

#### Example - Logback with enabled INCLUDE_MDC feature
Screenshot 2022-05-27 at 15 48 58

#### Example - Log4j with disabled INCLUDE_MDC feature
Screenshot 2022-05-27 at 15 57 54

#### Example - Log4j with enabled INCLUDE_MDC feature
Screenshot 2022-05-27 at 15 58 49

### Related Github Issue
The feature has been requested in [this issue #801](https://github.com/newrelic/newrelic-java-agent/issues/801).

### Testing
The agent includes a suite of tests that should be used to
verify your changes don't break existing functionality. These tests will run with
Github Actions when a pull request is made. More details on running the tests locally can be found
[here](https://github.com/newrelic/newrelic-java-agent/blob/main/CONTRIBUTING.md).

### Checks

[x] Are your contributions backwards compatible with relevant frameworks and APIs?
[ ] Does your code contain any breaking changes? Please describe. 
[ ] Does your code introduce any new dependencies? Please describe.

### CODE REVIEW

**Comment by StanlieK** [comment]
@kford-newrelic @jasonjkeller Hey guys! May I ask for a review, please? Thanks so much

**Comment by fieldju** [comment]
I want to plus one this PR.

I had to do [this](https://github.com/fieldju/newrelic-java-agent/commit/0f7dd2a64d1cbcb2f84736e80370bda08fb3b026) on my fork.

I implemented it slightly different (whitelist of MDC keys) but I'd be perfectly happy having all the MDC keys too.

I need this and #849 for this to be a feature complete logging solution for my microservices.

@StanlieK I'd prefer if my vars are *NOT* prefixed by the key `mdc`, so that when my go services which do not use MDC add their context variables, I can write a single query that gets the logs from the distributed microservices.

EX.

Our java services and go services add `PipelineId` and I want to just be able to search for the workflow id without having to do complicated queries.

On my fork I have implemented it like this

```
NEW_RELIC_OPTS="\
...
-Dnewrelic.config.application_logging.forward_mdc_properties.enabled=true \
-Dnewrelic.config.application_logging.forward_mdc_include_list=ActivityID,ActivityType,Attempt,RunID,TaskQueue,WorkerID,WorkflowID,WorkflowType,ExecutionID,PipelineID,Tenant,OrgID,EnvID
...
"
```

image

:point_up: `PipelineId` can show up as a root property

**Comment by fieldju** [comment]
I'd also like to say that I need this PR and #849 for this to be a feature complete logging solution for my microservices.

File: instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java [hunk 1]
```diff
@@ -52,11 +56,15 @@ public static void recordNewRelicLogEvent(LogEvent event) {
...
                 // Bail out and don't create a LogEvent if log message is empty
                 if (formattedMessage != null && !formattedMessage.isEmpty()) {
-                    HashMap<String, Object> logEventMap = new HashMap<>(DEFAULT_NUM_OF_LOG_EVENT_ATTRIBUTES);
+                    HashMap<String, Object> logEventMap = new HashMap<>(getDefaultNumOfLogEventAttributes(event.getContextData()));
 
                     logEventMap.put(MESSAGE, formattedMessage);
                     logEventMap.put(TIMESTAMP, event.getTimeMillis());
 
+                    if (isApplicationLoggingForwardingIncludeMdcEnabled() && event.getContextData() != null) {
+                        event.getContextData().forEach((key, value) -> logEventMap.put(MDC_ATTRIBUTE_PREFIX + key, value));
```
Comment (fieldju) [inline instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java:65]: I'd love to make the prefix optional.

When working with distributed microservices that are multi-language / framework its nice to have a single context key to search for.

I wouldn't want to have to write queries that are `WHERE mdc.foo = 'bar' AND foo = 'bar'` if I could just write `WHERE foo = 'bar'`

**Comment by StanlieK** [comment]
> I want to plus one this PR.
> 
> I ended up [implementing this slightly different](https://github.com/fieldju/newrelic-java-agent/commit/0f7dd2a64d1cbcb2f84736e80370bda08fb3b026) on my fork.
> 
> I did a whitelist of MDC keys but I'd be perfectly happy having all the MDC keys too.
> 
> @StanlieK I'd prefer if my vars are _NOT_ prefixed by the key `mdc`, so that when my go services which do not use MDC add their context variables, I can write a single query that gets the logs from the distributed microservices.
> 
> EX.
> 
> Our java services and go services add `PipelineId` and I want to just be able to search for the workflow id without having to do complicated queries.
> 
> On my fork I have implemented it like this
> 
> ```
> NEW_RELIC_OPTS="\
> ...
> -Dnewrelic.config.application_logging.forward_mdc_properties.enabled=true \
> -Dnewrelic.config.application_logging.forward_mdc_include_list=ActivityID,ActivityType,Attempt,RunID,TaskQueue,WorkerID,WorkflowID,WorkflowType,ExecutionID,PipelineID,Tenant,OrgID,EnvID
> ...
> "
> ```
> 
> image
> 
> ☝️ `PipelineId` can show up as a root property

@fieldju I like the idea of MDC properties not being prefixed. Or at least an optional prefix.

Also, I like the idea of whitelisted/blacklisted properties. It would be nice if all the properties will be allowed by default, as we have lots of properties from project to project. It would be madness to control each separately. In this case, it would be useful to have allowed and disallowed properties. If you'll set disallowed to `"*"`, it will exclude all the properties but allowed ones.

File: instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java (hunk 1 above)
Comment (StanlieK) [reply instrumentation/apache-log4j-2/src/main/java/com/nr/agent/instrumentation/log4j2/AgentUtil.java:65]: I completely agree with you, I'll incorporate it.

**Comment by fieldju** [comment]
@StanlieK Since they aren't too interesting in merging these PRs, we pivoted to an alternate way of shipping logs to New Relic that gives us total control.

We ended up making a custom logback layout and using the [New Relic Fluent Bit helm chart](https://github.com/newrelic/helm-charts/tree/f7a851795e71ee8228ed33e04a293aa21fe40448/charts/newrelic-logging)

You can see/use/copy that config [here in this gist](https://gist.github.com/fieldju/c4af82bd8663569476be2d5f575d19f8)

point_up this will take forward any k,v pair that is in the json blob including any [MDC kv pair](https://gist.github.com/fieldju/c4af82bd8663569476be2d5f575d19f8#file-armorycloudjsonlayout-java-L80).

image

**Comment by kford-newrelic** [comment]
@fieldju sorry that it's taking us a while to get to this PR - as you can probably imagine, it's not as simple as merging in this one PR - we need to make sure we have solutions that will work for as wide an audience as possible and that we can (hopefully) implement relatively consistently across other logging frameworks, etc.

The good new though is that we have adding MDC to our logs on our roadmap for the Jul-Sep quarter, so we expect to get to this reasonably soon!

**Comment by meiao** [comment]
@StanlieK writing to let you know that I started working on merging this PR.
Though for now, a lot of the work will be to get an agreement with the other language agents.

Because of that, this will not be in our upcoming 7.10 release, but most likely in our 7.11 release.

**Comment by meiao** [comment]
@StanlieK writing to let you know that I've started the work to integrate this into our agent.

For now  the work is to get an agreement with the other language agents so we have uniformity.
This process takes some time, but as soon as an agreement is reached, I will make the necessary modification and merge this PR.

**Comment by meiao** [comment]
This has been tested and reviewed in #1016

**Review by meiao — APPROVED** [review]

**Comment by mcflythekid** [comment]
Please help to change this document as it is not correct

- application_logging.forwarding.include_mdc.enabled to true
- NEW_RELIC_APPLICATION_LOGGING_FORWARDING_INCLUDE_MDC_ENABLED to true

**Comment by meiao** [comment]
@mcflythekid this PR was modified before it was added to the codebase.
Please refer to the documentation for the correct configuration:
https://docs.newrelic.com/docs/apm/agents/java-agent/configuration/java-agent-configuration-config-file/#log-context-data

**Pull request merged**